README.md
# MMM-LeagueOfLegends

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/).

A module to display your League Of Legends stats on your MagicMirror².

This is an example that uses this module 3 times: On the left hand side it displays the current rank in Solo and Flex Queue. In the middle it shows a match history of the latest 10 games.

![Preview](./Screenshot.png)

## Requirements

1. Riotgames Developer Profile
2. Riotgames Developer API-Key

Both of those are free to create, although it **might take a few days**, to get your application verified!

To use this module, you first need to sign up for the [Riotgames Developer Program](https://developer.riotgames.com/).

Next you need to register your Application with a `PERSONAL API KEY` by clicking the "Register Product"-Button on your Dashboard and selecting "Personal API Key". You then have to fillout a short from with information about the purpose of your "product" and submit this.

For me, it took a few days to get my application registered.

If your App is registered, you can head over to your "Apps" (by clicking on your username in the upper right corner and selecting "Apps") and choose your App on the left hand side. You should now see general Information about your App along with an "API-Key", looking something like this: `RGAPI-291b3053-432c-737l-28aj-416d3j21ag28` (not a real key, just an example!)

You are then ready to use this module.

## Installation

1. Navigate to your modules folder in a terminal:

```bash
cd modules
```

2. Clone this repository:

```bash
git clone https://github.com/JulianEgbert/MMM-LeagueOfLegends.git
```

### Updating

To update this module, navigate to the module folder (`cd modules/MMM-LeagueOfLegends`) and pull from the repository using:

```bash
git pull
```

## Using the module

To use this module, add the following configuration block to the modules array in the `config/config.js` file:

```js
var config = {
  modules: [
    { // Display your rank
      module: 'MMM-LeagueOfLegends',
      position: "middle_center",
      config: {
        summonerName: "YOUR_SUMMONER_NAME",
        region: "euw1",
        queueType: "RANKED_SOLO_5x5",
        apiKey: "YOUR_RIOT_API_KEY",
        displayElements: [
          {
            name: "tier",
            config: {
              hideDetailedRankInfo: true
            }
          },
          {
            name: "stats",
            config: {
              showHotStreak: true
            }
          }
        ],
      }
    },
    { // Display your match history
      module: 'MMM-LeagueOfLegends',
      position: "middle_center",
      config: {
        summonerName: "YOUR_SUMMONER_NAME",
        apiKey: "YOUR_RIOT_API_KEY",
        region: "euw1",
        matchRegion: "europe",
        displayElements: [
          {
            name: "history",
            config: {
              count: 10
            }
          },
        ],
      }
    },
  ]
}
```

## Configuration options

| Option           | Description
|----------------- |-----------
| `apiKey`         | *Required* Your own API-Key for the Riot API.
| `summonerName`   | *Required* Your summoner name.
| `displayElements`| *Required* The elements that should be displayed in this module. **Type**: Array of objects, see ["Display Elements"](##Display-Elements).
| `region`         | *Required* The region you are playing in (as stated in the riot api description). <br /> **Possible values**: `euw1` (default), `br1`, `eun1`, `jp1`, `kr`, `la1`, `la2`, `na1`, `oc1`, `ru`, `tr1`
| `matchRegion`         | *Required for history* The region you are playing in. It's different because the match API has different regions from the profile API. <br /> **Possible values**: `europe` (default), `americas`, `asia`
| `imageFolder`    | *Optional* The path to the folder with the tier icons. <br /> **Possible values**: `"emblems"` (default, new icons), `"tiers"` (old icons), or your custom folder name if you add one.
| `iconSize`       | *Optional* Size of the tier icon. <br /> **Default:** `256`
| `startDelay`     | *Optional* Time to wait before starting the module. Necessary if you have multiple modules because the number of API calls is time restricted. <br /> **Default:** `0`
| `updateInterval` | *Optional* How often the module reloads. <br /> **Default:** `300000` (5 minutes)
| `language`       | *Optional* Language used for the *time* display (everything else is currently english only!). <br /> **Default:** `en-EN`
| `queueType`      | *Optional* The queue you want to display your Elo from. <br /> **Possible values**: `"RANKED_SOLO_5x5"` (default, Solo-Queue), `"RANKED_FLEX_SR"` (Flex-Queue)
| `showOtherQueueIfNotFound` | *Optional* Whether the module should display another queue elo if specified queue is not found. <br />**Type:** Boolean, default: `false`


## Display Elements

Here is an overview of all the supported display elements. They are an object with `name` and `config` (*optional*) for configurations. If no config is provided, default values will be used.

| Option        | Description
|---------------|-----------
| `tier`        | Displays the icon of your rank. <br /> **Config**: `hideDetailedRankInfo`: *boolean* (**default** `false`), wether or not to display the tier name, division and LP.
| `stats`       | Displays the stats for this queue (Wins, Losses, Winrate). <br /> **Config**: `showHotStreak`: **Type** Boolean (**default** `false`), wether or not to display a flame icon, when the summoner has a hotstreak (provided by the riot API).
| `summoner`    | Displays the summoner name. <br /> **Config**: `showLevel`: *boolean* (**default** `false`), wether or not to display the level of the user.
| `clash`       | Displays the schedule of the next clash tournaments. <br /> **Config**: `count`: *number* (**default** `5`), number of upcoming tournaments to display. <br /> `fadeOut` (default `true`), wether the last two entries should fade out (lower opacity)
| `history`     | Displays the latest games with brief information. <br /> **Config**: `count` (default `5`): number of matches to display. <br /> `showTime` (default `true`): display date of game and game duration. <br /> `showStats` (default `true`): display stats of the game (kda and cs). <br /> `showQueue` (default `true`): display queue and win or loss. <br /> `showChampion` (default `true`): display the champion splashart. <br /> `iconSize` (default `64`): size of the champion icon in the history <br /> `csPerMinute` (default `true`): display cs per minute after cs score. <br /> `fadeOut` (default `true`), wether the last two entries should fade out (lower opacity)
| `live`        | Display your current live game. <br /> **Config**: `notIngameText` the text it displays when you are not ingame (default `"Currently not in a game"`).
| `friends`        | Display your friends that are ingame. <br /> **Config**: `friends` a list of your friends summoner Names (no default, example: `["Friend 1", "Friend 2", ...]`).
| `html`        | Display custom html content. <br /> **Config**: `html` the html content you want to display.


CHANGELOG.md
# MMM-LeagueOfLegends Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## [1.0.0] - Unreleased

First public release


node_helper.js
/* Magic Mirror
 * Node Helper: MMM-LeagueOfLegends
 *
 * By JulianEgbert
 * MIT Licensed.
 */

var NodeHelper = require("node_helper");

module.exports = NodeHelper.create({

	// Override socketNotificationReceived method.

	/* socketNotificationReceived(notification, payload)
	 * This method is called when a socket notification arrives.
	 *
	 * argument notification string - The identifier of the noitication.
	 * argument payload mixed - The payload of the notification.
	 */
	socketNotificationReceived: function(notification, payload) {
		if (notification === "MMM-LeagueOfLegends-NOTIFICATION_TEST") {
			console.log("Working notification system. Notification:", notification, "payload: ", payload);
			// Send notification
			this.sendNotificationTest(this.anotherFunction()); //Is possible send objects :)
		}
	},

	// Example function send notification test
	sendNotificationTest: function(payload) {
		this.sendSocketNotification("MMM-LeagueOfLegends-NOTIFICATION_TEST", payload);
	},

	// this you can create extra routes for your module
	extraRoutes: function() {
		var self = this;
		this.expressApp.get("/MMM-LeagueOfLegends/extra_route", function(req, res) {
			// call another function
			values = self.anotherFunction();
			res.send(values);
		});
	},

	// Test another function
	anotherFunction: function() {
		return {date: new Date()};
	}
});


MMM-LeagueOfLegends.js
/* Magic Mirror
 * Module: MMM-LeagueOfLegends
 *
 * By JulianEgbert
 * MIT Licensed.
 */

Module.register("MMM-LeagueOfLegends", {
	defaults: {
		updateInterval: 300000,
		startDelay: 0,
		region: "euw1",
		matchRegion: "europe",
		language: "en-EN",
		imageFolder: "emblems",
		queueType: "RANKED_SOLO_5x5",
		iconSize: 256,
		showOtherQueueIfNotFound: false,
		apiKey: "", // Required
		summonerName: "", // Required
		displayElements: [
			{
				name: "tier",
				config: {
					hideDetailedRankInfo: false
				}
			},
			{
				name: "stats",
				config: {
					showHotStreak: true
				}
			}
		],
	},

	requiresVersion: "2.1.0", // Required version of MagicMirror

	start: async function() {
		var self = this;
		this.loading = {done: 0, total: 0};
		this.summonerData = null;
		this.rankData = null;
		this.liveData = null;
		this.queueData = null;
		this.historyData = {};
		this.matchIds = null;
		this.version = null;
		this.championData = null;
		this.queues = null;
		this.friends = {};
		this.friendsData = {};
		this.clashData = null;

		//Flag for check if module is loaded
		this.loaded = false;

		await this.initData();
		// Allow module to load delayed, necessary for multiple modules and requests
		setTimeout(() => {
			this.updateData();
			// Schedule update timer.
			if (this.config.updateInterval > 0) {
				setInterval(async function() {
					self.updateData();
					self.updateDom();
				}, this.config.updateInterval);
			}
		}, this.config.startDelay);
	},

	initData: async function() {
		await this.getSummonerData();
		await this.getGameConstants();
		this.getInitialFriendsData();
		// this.updateData();
	},

	updateData: function() {
		this.loading = {done: 0, total: 0};
		const displayElements = this.config.displayElements.map((element) => element.name);
		if (displayElements.includes("tier") || displayElements.includes("stats")) {
			this.loading.total += 1;
			this.getRankData();
		}
		if (displayElements.includes("history")) {
			this.loading.total += 1;
			this.getHistoryData();
		}
		if (displayElements.includes("clash")) {
			this.getClashData();
		}
		if (displayElements.includes("live")) {
			this.getLiveData();
		}
		if (displayElements.includes("friends")) {
			this.getFriendsData();
		}
		if (this.loading.total === 0) {
			this.moduleLoaded();
		}
	},

	getSummonerData: async function() {
		var urlApi = `https://${this.config.region}.api.riotgames.com/lol/summoner/v4/summoners/by-name/${this.config.summonerName}?api_key=${this.config.apiKey}`;

		const response = await fetch(urlApi);
		const json = await response.json();
		this.summonerData = json;
	},

	getGameConstants: async function() {
		const versionUrl = "https://ddragon.leagueoflegends.com/api/versions.json";
		const queueUrl = "https://static.developer.riotgames.com/docs/lol/queues.json";

		const versionResponse = fetch(versionUrl);
		const queueResponse = fetch(queueUrl);

		await Promise.all([versionResponse, queueResponse]).then(async (results) => {
			const versions = await results[0].json()
			this.version = versions[0];
			this.queues = await results[1].json()
		});
		const championResponse = await fetch(`https://ddragon.leagueoflegends.com/cdn/${this.version}/data/en_US/champion.json`);
		this.championData = await championResponse.json();
		this.championData = this.championData.data; // the real data is stored in data.
	},

	getRankData: function() {
		if (!this.summonerData) {
			console.error(self.name, "No data for the summoner found");
		}
		var urlApi = `https://${this.config.region}.api.riotgames.com/lol/league/v4/entries/by-summoner/${this.summonerData.id}?api_key=${this.config.apiKey}`;

		this.sendRequest(urlApi, this.processRankData);
	},

	getLiveData: function() {
		if (!this.summonerData) {
			console.error(self.name, "No data for the summoner found");
		}
		var urlApi = `https://${this.config.region}.api.riotgames.com/lol/spectator/v4/active-games/by-summoner/${this.summonerData.id}?api_key=${this.config.apiKey}`;

		this.sendRequest(urlApi, (self, data) => {
			self.liveData = data;
			self.updateDom();
			self.ingameInterval = setInterval(() => self.updateDom(), 1000);
		}, (self, error) => {
			self.liveData = null;
			clearInterval(self.ingameInterval);
			self.updateDom();
		});
	},

	matchHistoryRequired: function() {
		const displayElements = this.config.displayElements.filter(element => typeof(element) === "object" && element.name === "history");
		return displayElements.length > 0;
	},

	matchHistoryConfig: function() {
		const historyElements = this.config.displayElements.filter(element => typeof(element) === "object" && element.name === "history");
		return historyElements[0].config;
	},

	getHistoryData: function() {
		if (!this.matchHistoryRequired()) {
			this.moduleLoaded();
			return;
		}
		if (!this.summonerData) {
			console.error(self.name, "No data for the summoner found");
		}
		const config = {
			count: 5
		};
		Object.assign(config, this.matchHistoryConfig());
		var urlApi = `https://${this.config.matchRegion}.api.riotgames.com/lol/match/v5/matches/by-puuid/${this.summonerData.puuid}/ids?count=${config.count}&api_key=${this.config.apiKey}`;
		this.sendRequest(urlApi, this.processHistoryData);
	},

	sendRequest: function(url, onSuccess, onError = null) {
		var self = this;
		var request = new XMLHttpRequest();
		request.open("GET", url, true);
		request.onreadystatechange = function() {
			if (this.readyState === 4) {
				if (this.status === 200) {
					onSuccess(self, JSON.parse(this.response));
				} else if (this.status === 401) {
					self.updateDom(self.config.animationSpeed);
					Log.error(self.name, this.status);
				} else {
					if (onError) {
						onError(self, this)
					} else {
						Log.error(self.name, `Could not load data from "${url}".`);
					}
				}
			}
		};
		try {
			request.send();
		}
		catch (e) {
			console.error(self.name, `Error loading data from "${url}"`);
		}
	},

	prepareQueueData: function() {
		if (!this.rankData || !Array.isArray(this.rankData) || this.rankData.length === 0) {
			return;
		}
		const queue = this.rankData.filter((queue) => queue.queueType === this.config.queueType);
		if (queue.length === 0 && this.config.showOtherQueueIfNotFound) { // Didn't find the queue specified in config:
			this.queueData = this.rankData[0];
			return;
		}
		this.queueData = queue[0];
	},

	getDom: function() {
		/* eslint-disable */
		let domBuilder = new DomBuilder(this, this.file(""));
		return domBuilder.getDom();
		/* eslint-enable */
	},

	getScripts: function() {
		return [
			this.file("helper/DomBuilder.js")
		];
	},

	getStyles: function () {
		return [
			"MMM-LeagueOfLegends.css",
		];
	},

	moduleLoaded() {
		this.loaded = true;
		this.updateDom(this.config.animationSpeed);
	},

	processRankData: function(self, data) {
		self.rankData = data;
		self.prepareQueueData();
		self.dataProcessed();
	},

	dataProcessed: function() {
		this.loading.done += 1;
		if (this.loading.total <= this.loading.done) {
			this.moduleLoaded();
		}
	},

	processHistoryData: function(self, data) {
		self.matchIds = data;
		data.forEach(matchId => {
			self.getMatchData(matchId);
		});
		self.dataProcessed();
	},

	getMatchData: function(matchId) {
		if (this.historyData && this.historyData[matchId])
			return;
		var urlApi = `https://${this.config.matchRegion}.api.riotgames.com/lol/match/v5/matches/${matchId}?api_key=${this.config.apiKey}`;
		this.sendRequest(urlApi, this.processMatchData);
	},

	getClashData: function() {
		const urlApi = `https://euw1.api.riotgames.com/lol/clash/v1/tournaments?api_key=${this.config.apiKey}`;
		this.sendRequest(urlApi, (self, data) => {
			self.clashData = data.sort((a,b) => { return a.schedule[0].startTime > b.schedule[0].startTime ? 1 : -1});
			self.updateDom();
		});
	},

	getInitialFriendsData: function() {
		const friendsModule = this.config.displayElements.filter((el) => el.name === "friends");
		if (friendsModule.length === 0)
			return;
		const friends = friendsModule[0].config.friends;
		friends.forEach((friend) => {
			var urlApi = `https://${this.config.region}.api.riotgames.com/lol/summoner/v4/summoners/by-name/${friend}?api_key=${this.config.apiKey}`;
			this.sendRequest(urlApi, (self, data) => {
				const name = data.name;
				self.friends[name] = data;
			});
		});
	},

	getFriendsData: function() {
		const friends = Object.keys(this.friends);
		friends.forEach((friend) => {
			this.getFriendData(this.friends[friend]);
		});
	},

	getFriendData: function(friend) {
		const url = `https://${this.config.region}.api.riotgames.com/lol/spectator/v4/active-games/by-summoner/${friend.id}?api_key=${this.config.apiKey}`;
		const name = friend.name;
		var self = this;
		var request = new XMLHttpRequest();
		request.open("GET", url, true);
		request.onreadystatechange = function() {
			if (this.readyState === 4) {
				if (this.status === 200) {
					self.updateFriendData(name, JSON.parse(this.response));
					// self.friendsData[name] = JSON.parse(this.response);
					// self.updateDom();
				} else if (this.status === 401) {
					self.updateDom(self.config.animationSpeed);
					Log.error(self.name, this.status);
				} else {
					self.updateFriendData(name, null);
					// self.friendsData[name] = null;
					// self.updateDom();
				}
			}
		};
		request.send();
	},

	updateFriendData: function(name, data) {
		this.friendsData[name] = data;
		const ingameFriends = Object.keys(this.friendsData).filter((f) => this.friendsData[f] !== null);
		if (ingameFriends.length > 0) {
			this.friendInterval = setInterval(() => this.updateDom(), 1000);
		} else {
			clearInterval(this.friendInterval);
			this.updateDom();
		}
	},

	processMatchData: function(self, data) {
		if (!self.historyData) {
			self.historyData = {};
		}
		self.historyData[data.metadata.matchId] = data;
		self.updateDom();
	}
});


helper/DomBuilder.js
class DomBuilder {

	constructor(self, pathPrefix) {
		this.pathPrefix = pathPrefix + '/';
		Object.assign(this, self);
  	}

	file(name) {
		return this.pathPrefix + name;
	}

  	getDom(){
		var wrapper = document.createElement("div");
		// If rankData is available
		if (!this.loaded) {
			wrapper.innerHTML = "LOADING";
			return wrapper;
		}
		this.config.displayElements.forEach(displayElement => {
			const name = displayElement.name.toLowerCase();
			switch(name) {
				case "tier":
					wrapper.appendChild(this.getTierDiv(displayElement.config));
					break;
				case "stats":
					wrapper.appendChild(this.getStatsDiv(displayElement.config));
					break;
				case "summoner":
					wrapper.appendChild(this.getSummonerDiv(displayElement.config));
					break;
				case "history":
					wrapper.appendChild(this.getHistoryDiv(displayElement.config));
					break;
				case "clash":
					wrapper.appendChild(this.getClashDiv(displayElement.config));
					break;
				case "live":
					wrapper.appendChild(this.getLiveDiv(displayElement.config));
					break;
				case "html":
					wrapper.innerHTML += displayElement.config.html;
					break;
				case "friends":
					wrapper.appendChild(this.getFriendsDiv(displayElement.config));
					break;
				default:
					wrapper.innerHTML += `<p> Can't find display element "${name}". <br> Please check your config for MMM-LeagueOfLegends </p>`
					break;
			}
		});
		if (wrapper.childNodes.length === 0) { // No data available, display information
			const errorWrapper = document.createElement("div");
			errorWrapper.innerHTML = `${this.name}: There is not data to display.`;
			wrapper.appendChild(errorWrapper);
		}
		return wrapper;
	}

	getMatchRow(matchId, config) {
		const match = this.historyData[matchId];
		const matchRow = document.createElement("tr");
		const summoner = this.getSummonerFromMatch(matchId);

		// Champion Icon
		if (config.showChampion) {
			const iconColumn = document.createElement("td");
			matchRow.appendChild(iconColumn);
			const championIcon = this.getChampionIcon(summoner.championName, config.iconSize);
			iconColumn.appendChild(championIcon);
		}

		// Game Info
		if (config.showQueue) {
			const infoColumn = document.createElement("td");
			matchRow.appendChild(infoColumn);
			const victoryLabel = document.createElement("div");
			victoryLabel.innerHTML = summoner.win ? "Victory" : "Defeat";
			victoryLabel.classList.add(summoner.win ? "victory" : "defeat");
			infoColumn.appendChild(victoryLabel);
			const gameLabel = document.createElement("div");
			gameLabel.innerHTML = this.getQueueFromId(match.info.queueId);
			infoColumn.appendChild(gameLabel);
		}

		// time
		if (config.showTime) {
			const timeColumn = document.createElement("td");
			matchRow.appendChild(timeColumn);
			const timeLabel = document.createElement("div");
			const options = { year: 'numeric', month: 'numeric', day: 'numeric' };
			const s = new Date(match.info.gameCreation).toLocaleDateString(this.config.language, options);
			timeLabel.innerHTML = s;
			timeColumn.appendChild(timeLabel);
			const durationLabel = document.createElement("div");
			durationLabel.innerHTML = this.getGameDurationString(match.info.gameDuration);
			timeColumn.appendChild(durationLabel);
		}

		// Player stats
		if (config.showStats) {
			const statColumn = document.createElement("td");
			matchRow.appendChild(statColumn);
			const kdaLabel = document.createElement("div");
			kdaLabel.innerHTML = `
				<span class="kills"> ${summoner.kills} </span>/
				<span class="deaths"> ${summoner.deaths} </span>/
				<span class="assists"> ${summoner.assists} </span>`;
			statColumn.appendChild(kdaLabel);
			const csLabel = document.createElement("div");
			csLabel.innerHTML = `${summoner.totalMinionsKilled} cs`;
			if (config.csPerMinute)
				csLabel.innerHTML += ` (${Math.round(summoner.totalMinionsKilled * 600 / match.info.gameDuration)/10}/min)`;
			statColumn.appendChild(csLabel);
		}
		return matchRow
	}

	getClashDiv(customConfig) {
		const wrapper = document.createElement("table");
		wrapper.setAttribute("class", "LoL-Clash small");
		if (!this.clashData)
			return wrapper;
		var config = {
			count: 5,
			fadeOut: true,
		};
		Object.assign(config, customConfig);
		if (config.fadeOut)
			wrapper.classList.add("fadeout");
		const events = this.clashData.slice(0, config.count);
		// wrapper.innerHTML = JSON.stringify(events);
		if (events.length === 0) {
			wrapper.innerHTML("There are no clash matches available.");
		} else {
			events.forEach(event => {
				wrapper.appendChild(this.getClashEventRow(event));
			});
		}
		return wrapper;
	}

	getClashEventRow(event) {
		const eventRow = document.createElement("tr");
		const iconColumn = document.createElement("td");
		iconColumn.setAttribute("class", "symbol align-right")
		eventRow.appendChild(iconColumn);
		iconColumn.setAttribute("class", "fas fa-trophy");
		const nameColumn = document.createElement("td");
		eventRow.appendChild(nameColumn);
		nameColumn.setAttribute("class", "title bright");
		const name = event.nameKey[0].toUpperCase() + event.nameKey.substring(1) + " Clash";
		const day = event.nameKeySecondary.match(/\d/)[0];
		nameColumn.innerHTML = `${name} (Day ${day})`;
		const dateColumn = document.createElement("td");
		dateColumn.setAttribute("class", "time light")
		eventRow.appendChild(dateColumn);
		const date = new Date(0);
		date.setUTCSeconds(event.schedule[0].startTime / 1000);
		const options = { year: 'numeric', month: 'long', day: 'numeric' };
		dateColumn.innerHTML = date.toLocaleDateString(this.config.language, options);
		return eventRow;
	}

	getStatsDiv(config) {
		const wrapper = document.createElement("div");
		wrapper.setAttribute("class", "LoL-Stats");
		// create the icon:
		var statsLabel = document.createElement("label");
		const q = this.queueData;
		if (!q) {
			return wrapper;
		}
		const winrate = Math.round(q.wins * 100 / (q.wins + q.losses));
		statsLabel.innerHTML = `${q.wins}W ${q.losses}L, ${winrate}% `;
		wrapper.appendChild(statsLabel);
		if (config && config.showHotStreak && q["hotStreak"]) {
			var hotStreakLabel = document.createElement("span");
			hotStreakLabel.setAttribute("class", "fa fa-fire");
			wrapper.appendChild(hotStreakLabel);
		}

		return wrapper;
	}

	getHistoryDiv(customConfig) {
		var config = {
			showTime: true,
			showStats: true,
			showQueue: true,
			showChampion: true,
			fadeOut: true,
			iconSize: 64,
			csPerMinute: true,
		};
		Object.assign(config, customConfig);
		const wrapper = document.createElement("table");
		wrapper.setAttribute("class", "LoL-Matchhistory small");
		if (config.fadeOut)
			wrapper.classList.add("fadeout");
		if (!this.historyData) {
			return wrapper;
		}
		this.matchIds.forEach(matchId => {
			const match = this.historyData[matchId];
			if (!match)
				return;
			wrapper.appendChild(this.getMatchRow(matchId, config));
		});
		if(wrapper.childNodes.length === 0) {
			wrapper.innerHTML = "There are no matches to display.";
		}
		return wrapper;
	}

	getSummonerDiv(config) {
		const wrapper = document.createElement("div");
		wrapper.setAttribute("class", "LoL-Summoner");
		var summonerNameLabel = document.createElement("label");
		summonerNameLabel.innerHTML = this.summonerData.name;
		if (config && config.showLevel) {
			summonerNameLabel.innerHTML += ` (Level ${this.summonerData.summonerLevel})`;
		}
		wrapper.appendChild(summonerNameLabel);

		return wrapper;
	}

	getTierDiv(config) {
		const wrapper = document.createElement("div");
		wrapper.setAttribute("class", "LoL-Tier");
		if (!this.queueData) {
			wrapper.innerHTML = "No data to display";
			return wrapper;
		}
		// create the icon:
		let tierIcon = document.createElement("img");
		const rankTier = this.getRankTier();
		tierIcon.src = this.file(`img/${this.config.imageFolder}/${rankTier}.png`);
		tierIcon.alt = rankTier;
		tierIcon.width = tierIcon.height = this.config.iconSize;
		wrapper.appendChild(tierIcon);

		if (config && config.hideDetailedRankInfo) {
			return wrapper;
		}
		// create the text with Tier, Division and LP if requested:
		const informationDiv = document.createElement("div");
		var tierLabel = document.createElement("label");
		const q = this.queueData;
		if (q && q.tier) {
			tierLabel.innerHTML = `${q.tier} ${q.rank} - ${q.leaguePoints} LP`;
		} else {
			tierLabel.innerHTML = "Unranked";
		}
		informationDiv.appendChild(tierLabel);
		wrapper.appendChild(informationDiv);

		return wrapper;
	}

	getRankTier() {
		return this.queueData ?(this.queueData.tier ? this.queueData.tier.toLowerCase() : "unranked") : "unranked";
	}

	getQueueFromId(id) {
		const queue = this.queues.filter(queue => queue.queueId === id)[0];
		const description = queue.description;
		if (!description)
			return queue.map.includes("Custom") ? "Custom" : "Other";
		const d = description.toLowerCase();
		if (d.includes("flex"))
			return "Flex";
		if (d.includes("solo"))
			return "Solo";
		if (d.includes("aram"))
			return "ARAM";
		if (d.includes("draft pick"))
			return "Normal draft";
		if (d.includes("bot"))
			return "Bot game";
		return description
	}

	getGameDurationString(timeInSeconds) {
		if (timeInSeconds <= 0 || timeInSeconds > 36000) // A large time also means that the game is still loading
			return "Loading...";
		const minutes = Math.floor(timeInSeconds / 60);
		const seconds = Math.floor(timeInSeconds % 60);
		return `${minutes < 10 ? "0"+minutes : minutes}:${seconds < 10 ? "0"+seconds : seconds}`;
	}

	getSummonerFromMatch(matchId) {
		const participants = this.historyData[matchId].info.participants;
		return participants.filter(p => p.puuid === this.summonerData.puuid)[0];
	}

	getChampionById(championId) {
		if (!this.championData) {
			return
		}
		const champions = Object.values(this.championData);
		const champion = champions.filter(champion => {return parseInt(champion.key) === championId})[0];
		return champion.id;
	}

	getLiveDiv(customConfig) {
		const config = {
			notIngameText: "Currently not in a game",
		};
		Object.assign(config, customConfig);
		const wrapper = document.createElement("table");
		wrapper.setAttribute("class", "small LoL-Livegame")
		const row = document.createElement("tr");
		wrapper.appendChild(row);
		if (this.liveData) {
			const iconCol = document.createElement("td");
			row.appendChild(iconCol);
			const player = this.liveData.participants.filter(player => player.summonerId === this.summonerData.id)[0];
			const queue = this.getQueueFromId(this.liveData.gameQueueConfigId);
			const champImage = this.getChampionIcon(this.getChampionById(player.championId), 64);
			iconCol.appendChild(champImage);
			const gameDetails = document.createElement("td");
			row.appendChild(gameDetails);
			const duration = (Date.now() - this.liveData.gameStartTime) / 1000;
			gameDetails.innerHTML = `<div> ${queue} </div> <div> ${this.getGameDurationString(duration)} </div>`;
		} else {
			wrapper.innerHTML = config.notIngameText;
		}
		return wrapper;
	}

	getChampionIcon(championName, size =64) {
		const championIcon = document.createElement("img");
		championIcon.src = `http://ddragon.leagueoflegends.com/cdn/${this.version}/img/champion/${championName}.png`;
		championIcon.width = championIcon.height = size;
		return championIcon;
	}

	getFriendsDiv(customConfig) {
		const table = document.createElement("table");
		Object.keys(this.friendsData).forEach((name) => {
			if (!this.friendsData[name])
				return
			table.appendChild(this.getFriendRow((name)));
		});
		if (table.childNodes.length === 0) {
			table.innerHTML = "<tr> No friends are ingame </tr>";
		}
		return table;
	}

	getFriendRow(name) {
		const matchInfo = this.friendsData[name];
		const player = matchInfo.participants.filter((p) => p.summonerName === name)[0];
		const queue = this.getQueueFromId(matchInfo.gameQueueConfigId);
		const duration = (Date.now() - matchInfo.gameStartTime) / 1000;
		const row = document.createElement("tr");
		const iconCol = document.createElement("td");
		iconCol.appendChild(this.getChampionIcon(this.getChampionById(player.championId), 32));
		row.appendChild(iconCol);
		row.innerHTML += `
		<td> ${name} </td>
		<td> ${queue} </td>
		<td> ${this.getGameDurationString(duration)} </td>
		`;
		return row
	}
}
